src/toxcore/dht/server/mod.rs

```diff
@@ -266,35 +274,35 @@ impl Server {
         self.initial_bootstrap.push(pn);
     }
 
-    /// Run initial bootstapping, It continues to sending NodesRequest to bootstrap nodes
-    /// until at least one entry inserted to Kbucket
-    /// and send NodesRequest to nodes in Kbucket if all nodes are discarded.
-    /// All nodes in Kbucket are discarded means that network is not available for a while.
-    fn run_recover_network_error(self) -> IoFuture<()> {
+    /// Run initial bootstrapping. It sends `NodesRequest` packet to bootstrap
+    /// nodes periodically if all nodes in Kbucket are discarded (including the
+    /// case when it's empty). It has to be an endless loop because we might
+    /// loose the network connection and thereby loose all nodes in Kbucket.
+    fn run_bootstrap_requests_sending(self) -> IoFuture<()> {
         let interval = Duration::from_secs(BOOTSTRAP_INTERVAL);
-
         let wakeups = Interval::new(Instant::now(), interval);
 
         let future = wakeups
             .map_err(|e| Error::new(ErrorKind::Other, format!("Bootstrap timer error: {:?}", e)))
             .for_each(move |_instant| {
                 trace!("Bootstrap wake up");
-                let close_nodes = self.close_nodes.read();
-                if close_nodes.is_empty() || close_nodes.is_all_discarded() {
-                    self.send_bootstrap_and_ping()
-                } else {
-                    Box::new(future::ok(()))
-                }
+                self.send_bootstrap_requests()
             });
 
         Box::new(future)
     }
 
-    /// Send NodesRequest to initial bootstapping nodes and nodes in Kbucket.
-    fn send_bootstrap_and_ping(&self) -> IoFuture<()> {
+    /// Check if all nodes in Kbucket are discarded (including the case when
+    /// it's empty) and if so then send `NodesRequest` packet to nodes from
+    /// initial bootstrap list and from Kbucket.
+    fn send_bootstrap_requests(&self) -> IoFuture<()> {
         let mut request_queue = self.request_queue.write();
         let close_nodes = self.close_nodes.read();
 
+        if !close_nodes.is_all_discarded() {
```

Comment:
```
fn main() {
    let x = [];
    let c = x.iter().all(|x: &i32| *x == 0);
    println!("c = {}", c);
}
ёёё

Reply:
are you sure we should not check `close_nodes.is_empty()`?

Reply:
Yes, we should send request if `close_nodes.is_empty` OR `close_nodes.is_all_discarded`. When close nodes is_empty the second condition will be always `true`.